### ISSUE TITLE
ckNew(): CkReductionMgr not constructed on all PEs

### ISSUE THREAD

**Issue by trquinn** [UIUC-PPL/charm#1647]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/1647*

---
After restarting from a checkpoint, and during CkIO operations, I am getting errors like:
```
ATP Stack walkback for Rank 52 starting:
  _start`start.S:113
  __libc_start_main`0x2aaaade4ac35
  main`0x2013cfd6
  ConverseInit`0x203db125
  ConverseRunPE`0x203daf21
  CsdScheduler`0x203dc5fc
  CsdScheduleForever`0x203dc2cb
  _processHandler(void*, CkCoreState*)`0x202f247b
  _invokeEntry(int, envelope*, void*)`0x202f051c
  CkDeliverMessageFree`0x202ecafc
  CkReductionMgr::AddToInactiveList(CkReductionInactiveMsg*)`ckreduction.C:848
  CkReductionMgr::checkAndAddToInactiveList(int, int)`stl_map.h:638
  std::pair<std::_Rb_tree_iterator<std::pair<int const, int> >, bool> std::_Rb_t
ree<int, std::pair<int const, int>, std::_Select1st<std::pair<int const, int> >,
 std::less<int>, std::allocator<std::pair<int const, int> > >::_M_insert_unique<
std::pair<int, int> >(std::pair<int, int>&&)`stl_tree.h:218
  std::local_Rb_tree_decrement(std::_Rb_tree_node_base*)`tree.cc:98
ATP Stack walkback for Rank 52 done
Process died with signal 11: 'Segmentation fault'
```
From examining the core file from the debugger, it is pretty clear that CkReductionMgr::inactiveList has not been initialized.  Adding CkPrintf()s to CkReductionMgr::CkReductionMgr() confirms that CkReductionMgr gets called on every PE except one.  In the above example, that one happens to be the first PE on Node 52, and so is the parent of all the other PEs on that node.  Also note that in this case the ckNew() call in CkIO is creating only one hundred or so elements on a thousand cores, so must PEs are inactive.

This is all with ChaNGa running on Blue Waters.  Example crashes are in /u/sciteam/trq/scratch/h82651mr and /u/sciteam/trq/scratch/h201824mr.

**Comment by stwhite91** [comment]
*Original date: 2017-08-01 20:52:32*

---
The following patch might affect this?

https://charm.cs.illinois.edu/gerrit/#/c/2538/

**Comment by trquinn** [comment]
*Original date: 2017-08-01 23:05:40*

---
That patch did not help. It looks like _processNodeBocInitMsg() doesn't get called at all for the relevant group.  Related: _processBocInitMsg() does not get called on the problem PE, but it does get called on rank 0 PEs on other nodes, as well as all other PEs.

**Comment by trquinn** [comment]
*Original date: 2017-08-02 02:37:27*

---
Could ONESIDED be the problem?  I stuck a CkPrintf in front of the CMK_ONSIDED_IMPL code in _processHandler() and the problem went away.

**Comment by stwhite91** [comment]
*Original date: 2017-08-02 16:33:49*

---
If ChaNGa isn't explicitly using the new (in v6.8.0) Zero-copy Send API, then no CMK_ONESIDED_IMPL code should ever execute.
To completely rule it out, you can try turning CMK_ONESIDED_IMPL off in src/arch/gni/conv-common.h

**Comment by trquinn** [comment]
*Original date: 2017-08-03 16:59:04*

---
A CkPrint in the RDMA code loop indicates that, as Sam suspected, the CMK_ONESIDED_IMPL is never executed.

It seems the most likely explanation is that the _processNodeBocInitMsg() on the Rank 0 PE on a node is not being called before an AddToInactiveList() message from a Rank > 0 PE is being processed.  This message gets generated as part of Group creation.  I've also seen crashes in CkArray::initDone().  This gets generated as part of CkArray creation since opts.staticInsertion is true for the CkIO ckNew().

Is it guaranteed that _processNodeBocInitMsg() on the Rank 0 PE gets called before these messages from other ranks get handled?

**Comment by trquinn** [comment]
*Original date: 2017-08-05 00:02:25*

---
I've now learned about _lookupGroupAndBufferIfNotThere().  It checks if there is a NULL entry in _groupTable, and queues the message if so.  The problem seems that for the CKIO ckNew(), a group gets created for which there is a non-NULL entry in _groupTable: the entry is set while reading the checkpoint.  However this entry gets initialized during restart, so it is non-NULL before the BocInitMsg arrives, and no queuing happens.  Furthermore, the epIdx for the CkCreateLocalGroup() call is different: during the CKIO ckNew() it is "CkArray(const CkArrayOptions &opts, const CkMarshalledMessage &ctorMsg)", but during the restart it was "CkMulticastMgr(CkMigrateMessage* impl_msg)".  This kind of explains why the object is somewhat sensible (e.g., thisgroup is assigned correctly) but stuff specific to CkReductionMgr is garbage.

So it looks like a checkpoint/restart problem: it is putting semi-garbage into the _groupTable.

**Comment by stwhite91** [comment]
*Original date: 2017-08-08 21:29:45*

---
Redmine #1652 is believed to be the root cause of this issue, and it's fix has been merged into 6.8.0.
